src/lib/peers/peer-connection/PeerConnection.ts

```diff
@@ -0,0 +1,239 @@
+      );
+      const peerClient = new PeerClient(connectedAddress.toString(), this.credentials);
+
+      this.connected = true;
+      return peerClient;
+    } else if (!this.getPeerInfo().relayPublicKey) {
+      throw Error('peer does not have relay public key specified');
+    } else {
+      throw Error('peer is already connected');
+    }
+  }
+
+  private async connectRelay(): Promise<PeerClient> {
+    // try to relay to peer via relay peer
+    if (!this.connected && this.getPeerInfo().relayPublicKey) {
+      // turn relay
+      // connect to relay and ask it to create a relay
+      const connectedAddress = await this.peerManager.turnClient.requestPeerConnection(
+        this.getPeerInfo().publicKey,
+        this.getPeerInfo().relayPublicKey!,
+      );
+      const peerClient = new PeerClient(connectedAddress.toString(), this.credentials);
+      this.connected = true;
+      return peerClient;
+    } else if (!this.getPeerInfo().relayPublicKey) {
+      throw Error('peer does not have relay public key specified');
+    } else {
+      throw Error('peer is already connected');
+    }
+  }
+
+  async connectFirstChannel() {
+    return await new Promise<PeerClient>((resolve, reject) => {
+      const promiseList = [this.connectDirectly(), this.connectHolePunch(), this.connectRelay()];
+
+      const errorList: Error[] = [];
+      for (const promise of promiseList) {
+        promise
+          .then((p) => {
+            resolve(p);
+          })
+          .catch((error) => null);
+
+        promise.catch((error) => {
+          errorList.push(error);
+          // check if all have failed
+          if (errorList.length == promiseList.length) {
+            reject(errorList);
+          }
+        });
+      }
+    });
+  }
+
+  private async connect(): Promise<void> {
+    // connect if not already connected
+    if (!this.connected) {
+      try {
+        this.peerClient = await this.connectFirstChannel();
+      } catch (error) {
+        console.log(error);
+      }
+    }
+
+    // try a ping
+    if (this.connected && (await this.sendPingRequest(5000))) {
+      return;
+    } else {
+      this.connected = false;
+      // still not connected
+      throw Error('could not connect to peer');
+    }
+  }
+
+  private getPeerInfo() {
+    if (!this.peerManager.hasPeer(this.publicKey)) {
+      throw Error('peer does not exist in peer store');
+    }
+    return this.peerManager.getPeer(this.publicKey)!;
+  }
+
+  private async sendPingRequest(timeout?: number): Promise<boolean> {
+    // eslint-disable-next-line
+    return await new Promise<boolean>(async (resolve, reject) => {
+      try {
+        if (timeout) {
+          setTimeout(() => reject('ping timed out'), timeout);
+        }
+
+        const challenge = randomBytes(16).toString('base64');
+
+        // encode request
+        const peerRequest = await this.encodeRequest(SubServiceType.PING_PEER, stringToProtobuf(challenge));
```

Comment:
Are you aware of HMAC messages? I think we should use this in our protobuf. Even JWT is nice. Could work nicely with the OAuth work.

Reply:
I think TLS might use HMAC under the hood? but it's a good idea to confirm this.

Reply:
yeah it looks like it:
https://support.venafi.com/hc/en-us/articles/215911297-Info-What-is-the-difference-between-SSL-and-TLS-

Reply:
I think you meant more for the initial authentication right? thats where JWT could fit in I guess